model/DataStore/MetaDataDeliverySyncTask.php

```diff
@@ -0,0 +1,171 @@
+    {
+        return __CLASS__;
+    }
+
+    /**
+     * @throws InvalidServiceManagerException
+     */
+    private function getQueueDispatcher(): ConfigurableService
+    {
+        return $this->getServiceManager()->get(QueueDispatcher::SERVICE_ID);
+    }
+
+    /**
+     * @param $params
+     * @throws InvalidServiceManagerException
+     */
+    private function requeueTask($params): void
+    {
+        /** @var QueueDispatcher $queueDispatcher */
+        $queueDispatcher = $this->getQueueDispatcher();
+        $queueDispatcher->createTask(
+            $this,
+            $params,
+            __('Continue try to sync GCP of delivery "%s".', $params['deliveryId'])
+        );
+    }
+
+    private function getPersistDataService(): PersistDataService
+    {
+        return $this->getServiceLocator()->get(PersistDataService::class);
+    }
+
+    /**
+     * @throws common_exception_Error
+     * @throws core_kernel_persistence_Exception
+     */
+    private function prepareData($params)
+    {
+        $params['count'] = $params['count'] ?? 0;
+        if (!isset($params['deliveryMetaData'], $params['testMetaData'], $params['testUri'], $params['itemMetaData'])) {
+            $compiler = $this->getMetaDataCompiler();
+            //DeliveryMetaData
+            $deliveryResource = new core_kernel_classes_Resource($params['deliveryId']);
+            $params['deliveryMetaData'] = $compiler->compile($deliveryResource);
+            //test MetaData
+            $test = $this->getTest($deliveryResource);
+            $params['testUri'] = $this->getTestUri($deliveryResource);
+            $params['testMetaData'] = $compiler->compile($test);
+            //Item MetaData
+            $params['itemMetaData'] = $this->getItemMetaData($test, $compiler);
+        }
+
+        return $params;
+    }
+    private function getItemMetaData(core_kernel_classes_Resource $test, ResourceJsonMetadataCompiler $compiler): array
+    {
+        /** @var taoQtiTest_models_classes_QtiTestService $testService */
+        $testService = $this->getServiceLocator()->get(taoQtiTest_models_classes_QtiTestService::class);
+        $items = $testService->getItems($test);
+        $itemMetaData = [];
+        foreach ($items as $item) {
+            $itemMetaData[] = $compiler->compile($item);
+        }
+
+        return $itemMetaData;
+    }
+
+    /**
+     * @throws common_exception_Error
+     * @throws core_kernel_persistence_Exception
+     */
+    private function getTest(core_kernel_classes_Resource $deliveryResource): core_kernel_classes_Resource
+    {
+        $testUri = $this->getTestUri($deliveryResource);
+
+        return new core_kernel_classes_Resource($testUri);
+    }
+
+    /**
+     * @throws core_kernel_persistence_Exception
+     */
+    private function getTestUri(core_kernel_classes_Resource $deliveryResource): ?string
+    {
+        $testProperty = new core_kernel_classes_Property(DeliveryAssemblyService::PROPERTY_ORIGIN);
```

Comment:
You should use `$this->getResouce()` instead
```suggestion
        $testProperty = $this-getProperty(DeliveryAssemblyService::PROPERTY_ORIGIN);
```

Reply:
even with OntologyAwareTrait added this is not found as a function 
can you provide more info on that?

Reply:
`oat\generis\model\OntologyAwareTrait::getProperty()`

Reply:
I checked the trait and it is there. Maybe you did not import the trait? I think the point is to centralize the object encapsulation inside the trait only.

Reply:
done